### DISCUSSION TITLE
Are relationship options supposed to override loader options?

### DISCUSSION THREAD

**Discussion by lunalucadou** [sqlalchemy/sqlalchemy#11572] [Usage Questions]
I am investigating some performance regressions in our application and have tracked them down to a seemingly innocuous change: adding eager loading to a declarative mapping relationship.

Setup:

```python
class Person(Base):
    __tablename__ = "person"
    id: Mapped[int] = mapped_column(primary_key=True)
    # ...
    addresses: Mapped[List["Address"]] = relationship(
        Address,
        back_populates="person",
        lazy="selectin",
        order_by=Address.id,
    )
    phone_numbers: Mapped[List["PhoneNumber"]] = relationship(
        PhoneNumber,
        back_populates="person",
        lazy="selectin",
        order_by=PhoneNumber.id,
    )
    # ...

class Address(Base):
    __tablename__ = "address"
    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey("person.id"))
    line_1: Mapped[str]
    # ...
    address_type: Mapped[str]
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="addresses",
        lazy="selectin",
        uselist=False,
    )

class PhoneNumber(Base):
    __tablename__ = "phone_number"
    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey("person.id"))
    phone_number: Mapped[str]
    # ...
    phone_type: Mapped[str]
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="phone_numbers",
        lazy="selectin",
        uselist=False,
    )

# etc
```

Part of our application creates bulk exports. To save time (and memory), if a given export does not request all 15 subresources available, the only attributes we fetch for those non-requested subresources are their IDs and who controls them:

```python
loader_options: list = []
if Address.__tablename__ not in requested_resources:
    loader_options.append(
        selectinload(Person.addresses).options(
            load_only(Address.id, Address.owner)
        )
    )
if PhoneNumber.__tablename__ not in requested_resources:
    loader_options.append(
        selectinload(Person.phone_numbers).options(
            load_only(PhoneNumber.id, PhoneNumber.owner)
        )
    )
# etc
return session.scalars(stmt.options(*loader_options)).all()
```

This setup worked as expected. The problem is we've been redoing how we approach permissions, which also requires changing how we access a few subresources that have restricted data (e.g. non-work addresses and phones).

Because of the change to how we access the subresources, we had to add `join_depth=2` to the `PhoneNumber.person` and `Address.person` relationships. A seemingly innocuous change, except it is overriding the loader options for our bulk exports, resulting in all attributes on all subresources being loaded all the time, causing some exports to run out of memory and get killed. Unfortunately, we hadn't thought to write a test that made sure the attributes remained unloaded (just that the ID and owner were loaded), so we only found out after it hit production.

I can't seem to find a workaround for this, so I'm wondering:

* Are mapper relationship options supposed to override loader options?
* If so, is there a way to give loader options precedence?
* If not, is what I'm seeing intended behavior? And are there any workarounds?
* Is there any documentation on what takes precedence/order of evaluation?

**Comment by zzzeek** [comment]
hi -

It's a bit unusual to use join_depth on relationships that aren't immediately self-referential, that is, it's not something like `Person->Person`.   I see here that there can be traversals along backrefs like Person->PhoneNumber->Person but that usually doesnt need join_depth because a load like PhoneNumber->Person is many-to-one, and the related Person is pulled from the local identity map without any SQL emitted.

so I am more interested in seeing what the use case for join_depth is here first.

Overall, a loader option always takes precedence over a mapper-level loading scheme.     But I see the option is `selectinload(Person.phone_numbers).options(load_only(PhoneNumber.id, PhoneNumber.owner))` , I can't see what "owner" is, is that the `Person`?  This wouldnt get in the way of what normally is traversing from Person->PhoneNumber->Person.   if the selectinload() option then went onto include `lazyload()` for the next link in the chain, that should in theory cancel the mapper-level eager loaders.   but this looks like a very esoteric example so I would need see an exact reproduction script to know what happens.

**Reply by lunalucadou** [reply]
Our previous permissions process was basically a blacklist (strip out information the requestor doesn't have permission to view after it's already been retrieved), and we've been working to make more like a whitelist (make it so the application doesn't even touch information the requestor isn't allowed to view).

To accomplish this, we subclassed our Person mapping with a custom join condition that only includes information requestors at that tier are allowed to see:

```python
class PersonInternal(Person):
    addresses: Mapped[List["Address"]] = relationship(
        Address,
        back_populates="person",
        lazy="selectin",
        order_by=Address.id,
        primaryjoin="and_(PersonInternal.id == Address.person_id, Address.address_type == 'office')",
    )
    # repeat for phone_numbers & email_addresses
```

Since we are going to take on more sensitive information in the future with higher permission tiers, this will be easier to maintain going forward and less error-prone - a requestor's permission level sets the mapping its queries will be executed with, preventing unauthorized disclosures with much less code. (The bulk request system also uses this, but I wrote `Person.addresses` since that's less likely to cause confusion than `params.mapping.addresses`.)

However, this approach does require us to interact with sub-resources in a different manner. Previously, we employed the visitor pattern for all sub-resource interactions, grabbing sub-resources and using `getattr` to interact with them to determine which should stay. (It's not pretty, but interacting with them via the ORM wasn't giving us much headroom in our SLO budget.) But with these changes, we have to do `person.addresses` instead of grabbing all the addresses of the person and then figuring out which ones the requestor is allowed to see.

And the reason we had to add `join_depth=2` is because, if you want to just grab an address from our API, we also generate a resource link back to the base person. Previously, this was filling in our URL template with `address.person.id`, but now it's reaching for `person.address.person.id`, which results in a `DetatchedInstanceError` because SQLAlchemy didn't recursively load the relationship.

`[a.person = person for a in person.addresses]` allows our code to run with no `join_depth` specified, but it results in duplicate addresses in the ORM object. (The duplicate addresses aren't persisted, but having to de-duplicate the output is far from an ideal solution.) A deeper refactor than what was already performed is arguably in order, but I'm just trying to get things back to green for now.

Given all this, is `join_depth=2` in a mapper relationship overriding the loader options expected behavior?

> Person->PhoneNumber->Person but that usually doesnt need join_depth because a load like PhoneNumber->Person is many-to-one, and the related Person is pulled from the local identity map without any SQL emitted.

So `person.addresses[0].person` should work with `lazy="selectin"`, no `join_depth` required? [This documentation](https://docs.sqlalchemy.org/en/20/orm/self_referential.html#self-referential-eager-loading) would seem to indicate otherwise.

> I can't see what "owner" is, is that the Person?

No, it's a multi-tenant application thing. Whether a person returns a 200 or 404 and what that person even looks like depends on who's asking (the records are unified under the hood but we have to isolate and only fill in the information that each campus "owns" because FERPA).

**Reply by lunalucadou** [reply]
While I am still curious if relationship options overriding loader options is intended behavior, I am happy to report I have found a workaround: more subclasses. Each sub-resource with a relationship specifying a `join_depth` gets a subclass (whose sole member is a copy of the relationship, minus the `join_depth`), and then I reference them in a subclass of Person (whose only members are copies of Person's relationships, but referencing the subclasses of the other sub-resources), like so:

```python
class Person(Base):
    __tablename__ = "person"
    id: Mapped[int] = mapped_column(primary_key=True)
    # ...
    addresses: Mapped[List["Address"]] = relationship(
        Address,
        back_populates="person",
        lazy="selectin",
        order_by=Address.id,
    )
    phone_numbers: Mapped[List["PhoneNumber"]] = relationship(
        PhoneNumber,
        back_populates="person",
        lazy="selectin",
        order_by=PhoneNumber.id,
    )
    # ...

class ExportablePerson(Person):
    addresses: Mapped[List["ExportableAddress"]] = relationship(
        ExportableAddress,
        back_populates="person",
        lazy="selectin",
        order_by=ExportableAddress.id,
    )
    phone_numbers: Mapped[List["ExportablePhoneNumber"]] = relationship(
        ExportablePhoneNumber,
        back_populates="person",
        lazy="selectin",
        order_by=ExportablePhoneNumber.id,
    )

class Address(Base):
    __tablename__ = "address"
    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey("person.id"))
    line_1: Mapped[str]
    # ...
    address_type: Mapped[str]
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="addresses",
        lazy="selectin",
        uselist=False,
        join_depth=2,
    )

class ExportableAddress(Address):
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="addresses",
        lazy="selectin",
        uselist=False,
    )

class PhoneNumber(Base):
    __tablename__ = "phone_number"
    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey("person.id"))
    phone_number: Mapped[str]
    # ...
    phone_type: Mapped[str]
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="phone_numbers",
        lazy="selectin",
        uselist=False,
        join_depth=2,
    )

class ExportablePhoneNumber(PhoneNumber):
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="phone_numbers",
        lazy="selectin",
        uselist=False,
    )
```

I can scale this quite easily with our permissions model as well (i.e. I can create a `ExportablePersonInternal(PersonInternal)` subclass that is a copy of `PersonInternal`, with the classes specified in the relationships switched to their `Exportable` counterparts).

After refactoring the exporter service to make use of the new mappings, the application is back to normal.

**Reply by zzzeek** [reply]
okay, so, I was looking for a specific example, which I've reproduced below based on your second example illustrating how you are using join_depth (which is what I asked to see).    So I can show you how to use options to override those, which is to use lazyload() for that path.  load_only() does not affect relationships.   You might want to use this rather than that more complicated mapping that you have since you dont have polymorphic identity for ExportableXYZ you can run into a situation where you have the same row in your Session more than once via the two different classes.   using the correct loader options would have a more local effect.  


```py
from __future__ import annotations

from typing import List

from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import ForeignKey
from sqlalchemy import select
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import lazyload
from sqlalchemy.orm import load_only
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship
from sqlalchemy.orm import selectinload
from sqlalchemy.orm import Session


class Base(DeclarativeBase):
    pass


class Person(Base):
    __tablename__ = "person"
    id: Mapped[int] = mapped_column(primary_key=True)
    # ...
    addresses: Mapped[List["Address"]] = relationship(
        back_populates="person",
        lazy="selectin",
        order_by="Address.id",
    )
    phone_numbers: Mapped[List["PhoneNumber"]] = relationship(
        back_populates="person",
        lazy="selectin",
        order_by="PhoneNumber.id",
    )
    # ...


class Address(Base):
    __tablename__ = "address"
    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey("person.id"))
    line_1: Mapped[str]
    # ...
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="addresses",
        lazy="selectin",
        uselist=False,
        join_depth=2,
    )


class PhoneNumber(Base):
    __tablename__ = "phone_number"
    id: Mapped[int] = mapped_column(primary_key=True)
    person_id: Mapped[int] = mapped_column(ForeignKey("person.id"))
    phone_number: Mapped[str]
    # ...
    person: Mapped["Person"] = relationship(
        "Person",
        back_populates="phone_numbers",
        lazy="selectin",
        uselist=False,
        join_depth=2,
    )


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)
s = Session(e)

p1 = Person()
a1 = Address(line_1="a1 line1", person=p1)
pn1 = PhoneNumber(phone_number="pn1", person=p1)


s.add_all([p1, a1, pn1])
s.commit()


def ensure_join_depth_is_two():
    with Session(e) as s:
        stmt = select(Person)

        ps = s.scalars(stmt).all()

    # session is closed.  this traversal fails without the extra join
    # depth.
    for p in ps:
        for ad1 in p.addresses:
            for ph in ad1.person.phone_numbers:
                ph.person


def ensure_options_override_join_depth():
    with e.connect() as conn, Session(conn) as s:

        stmt_count = 0

        @event.listens_for(conn, "before_cursor_execute")
        def count_statements(*arg, **kw):
            nonlocal stmt_count
            stmt_count += 1

        # we want three queries.  the SELECT for Person, SELECT for addresses,
        # SELECT for phone_numbers.   so we turn off the extra loads using
        # lazyload.   load_only() only affects columns, not relationships
        stmt = select(Person).options(
            selectinload(Person.addresses).options(
                lazyload(Address.person), load_only(Address.id)
            ),
            selectinload(Person.phone_numbers).options(
                lazyload(PhoneNumber.person), load_only(PhoneNumber.id)
            ),
        )

        ps = s.scalars(stmt).all()

        assert stmt_count == 3


ensure_join_depth_is_two()

ensure_options_override_join_depth()

```

**Reply by lunalucadou** [reply]
I like the simplicity of lazyload (as opposed to the inheritance method), and at first it seemed to be a much simpler solution to the problem, but upon further inspection, it does re-introduce the undesired behavior in some cases.

* When loading no sub-resources (just the base Person), no problem - only loads the ID for sub-resources.
* When only loading sub-resources that do not have `join_depth` set on their relationship (e.g. Person.jobs), no problem - loads everything for Person.jobs, and only ID for all other sub-resources.
* But when loading sub-resources with `join_depth` set on their relationship (e.g. Person.addresses), whether in isolation or in combination with sub-resources that do not have `join_depth` set, it causes everything to get loaded for all sub-resources.

I've also tried a combination of the two methods (Person.addresses and Person.phone_numbers get selectinload with lazyload + load_only options, all others just get selectinload with load_only options) but the result is the same.

If this is intended behavior, we will probably have to stick with the inheritance method, messy though it may be.

> since you dont have polymorphic identity for ExportableXYZ you can run into a situation where you have the same row in your Session more than once via the two different classes

It is one codebase but the bulk exports and REST API use separate main methods, so although I can see how it could cause problems, I'm not sure that's much of a concern for our usage patterns (it's mostly read-only data that's populated by other processes).

**Reply by zzzeek** [reply]
>But when loading sub-resources with join_depth set on their relationship (e.g. Person.addresses), whether in isolation or in combination with sub-resources that do not have join_depth set, it causes everything to get loaded for all sub-resources.

if you can modify my self-contained script to show what you mean, I can continue ensuring there are no bugs in SQLAlchemy and/or fix them if there are.   loader options should give you complete control within their scope of operation.


Also I still suspect you'd be better off not using join_depth here and using the selectinload() options you otherwise want to get for that for those specific queries where you need that behavior.  it's easier to add options to provide for eager loading as needed rather than turning it on and providing options to turn it off.
